nomad/operator_endpoint.go

```diff
@@ -228,6 +228,134 @@ REMOVE:
+		return err
+	}
+	op.srv.MeasureRPCRate("operator", structs.RateMetricWrite, args)
+	if authErr != nil {
+		return structs.ErrPermissionDenied
+	}
+
+	// Check management permissions
+	if aclObj, err := op.srv.ResolveACL(args); err != nil {
+		return err
+	} else if aclObj != nil && !aclObj.IsManagement() {
+		return structs.ErrPermissionDenied
+	}
+
+	return op.doTransfer(args)
+}
+
+func (op *Operator) doTransfer(args any) error {
+	var id raft.ServerID
+	var addr raft.ServerAddress
+	var kind, testedVal string // to be used in making error strings
+
+	if lAddr, lID := op.srv.raft.LeaderWithID(); id == lID && addr == lAddr {
+		op.logger.Debug("leadership transfer to current leader is a no-op")
+		return nil
+	}
+
+	minRaftProtocol, err := op.srv.MinRaftProtocol()
+	if err != nil {
+		return err
+	}
+
+	// TransferLeadership is not supported until Raft protocol v3 or greater.
+	if minRaftProtocol < 3 {
+		op.logger.Warn("unsupported minimum common raft protocol version", "required", "3", "current", minRaftProtocol)
+		return fmt.Errorf("unsupported minimum common raft protocol version")
+	}
+
+	// This is in a scope so that future will go out of scope once we're done
+	// checking the configuration, enabling us to reuse the name later.
+	{
+		// Get the raft configuration
+		future := op.srv.raft.GetConfiguration()
+		if err := future.Error(); err != nil {
+			return err
+		}
+
+		// Since this is an operation designed for humans to use, we will return
+		// an error if the supplied id isn't among the peers since it's
+		// likely a mistake.
+		switch tArg := args.(type) {
+		case *structs.RaftPeerByAddressRequest:
+			kind = "address"
+			testedVal = string(tArg.Address)
+			for _, s := range future.Configuration().Servers {
+				if s.Address == tArg.Address {
+					id = s.ID
+					addr = s.Address
+					goto TRANSFER
+				}
+			}
+		case *structs.RaftPeerByIDRequest:
+			kind = "id"
+			testedVal = string(tArg.ID)
+			for _, s := range future.Configuration().Servers {
+				if s.ID == tArg.ID {
+					id = s.ID
+					addr = s.Address
+					goto TRANSFER
+				}
+			}
+		default:
+			// This should never happen, since this function's callers ensure that the
+			// type is either a RaftPeerByAddressRequest or RaftPeerByIDRequest via
+			// their argument types; however, it can't hurt to be defensive for the
+			// future and it feels better to make an error than to panic for missing
+			// (but non-critical) functionality.
+			op.logger.Error("doTransfer: invalid argument", "type", fmt.Sprintf("%T", args))
+			return fmt.Errorf("doTransfer: invalid argument type (%T)", args)
+		}
+		return fmt.Errorf("%s %q was not found in the Raft configuration",
+			kind, testedVal)
+	}
+
+TRANSFER:
+	log := op.logger.With("peer_id", id, "peer_addr", addr)
+	if err = op.srv.leadershipTransferToServer(id, addr); err != nil {
+		log.Error("failed transferring leadership", "error", err)
+		return err
+	}
+
+	log.Info("transferred leadership")
```

Comment:
Might be nice to have the `from` and `to` in here if we can. It'll make for good post-incident forensics. Alternately, we have the `log.Info` already in `leadershipTransferToServer` so it's redundant here and we could have `from` and `to` fields there.